modules/internal/String.chpl

```diff
@@ -1587,6 +1587,32 @@ module String {
       return _join(S);
     }
 
+    /*
+      :args listToJoin: list contaning elements whose string values will be concatenated
+      .. code-block:: chapel
+      var x = "|".join([1,2,3])
+      writeln(x); //prints "a|10|d"
+    
+    */
+
+    proc join(listToJoin: list) : string {
+      return _join(listToJoin);
```

Comment:
I think we might be doing a bit too much work here. Recall that there is an overload of `join` that takes an iterator (the type _iteratorRecord).

This means that we can pass it iterators from arbitrary types, as long as they have a `these` method defined for them.

Luckily, the `list` type happens to have a `these` method defined for it...